src/nnfusion/engine/pass/graph/autodiff/roll.cpp

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "backward_registry.hpp"
+
+REGISTER_BACKWARD_TRANSLATOR(Roll).translator(
+    [](std::shared_ptr<GNode> forward_node,
+       const GNodeIndexVector& outputs_grad,
+       std::shared_ptr<nnfusion::graph::Graph> graph) -> GNodeIndexVector {
+        NNFUSION_CHECK(outputs_grad.size() == 1) << "Roll have only 1 output, but "
+                                                 << outputs_grad.size() << " outputs_grad provided";
+        // y = roll(x, shifts, dims)
+        // x_grad = roll(f_grad, fmap(reverse_list(shifts), [](int64_t i){return -i;}), reverse_list(dims))
+        auto generic_op = static_pointer_cast<nnfusion::op::GenericOp>(forward_node->get_op_ptr());
+        std::vector<int> shifts = generic_op->localOpConfig.getRoot()["shifts"];
+        std::vector<size_t> dims = generic_op->localOpConfig.getRoot()["dims"];
+
+        auto f_shifts = std::vector<int>(shifts.size());
+        std::transform(
+            shifts.rbegin(), shifts.rend(), f_shifts.begin(), [](int i) -> int { return -i; });
+        auto f_dims = dims;
```

Comment:
>auto f_dims = dims; [](http://example.com/codeflow?start=8&length=19)

❔[Question]

Why do we need this? Why not run the following operations on dims instead of f_dims to avoid an extra copy?

Reply:
You are right, changed to run on "dims" instead of "f_dim"